components/raftstore/src/store/fsm/peer.rs

```diff
@@ -3248,7 +3249,13 @@ where
             }
         }
         let allow_replica_read = read_only && msg.get_header().get_replica_read();
-        if !(self.fsm.peer.is_leader() || is_read_index_request || allow_replica_read) {
+        let flags = WriteBatchFlags::from_bits_truncate(msg.get_header().get_flags());
+        let allow_stale_read = read_only && flags.contains(WriteBatchFlags::STALE_READ);
+        if !(self.fsm.peer.is_leader()
```

Comment:
It's hard to read, how about `!is_leader && !is_read_index_request && !allow_replica_read && !allow_stale_read`?